### PR TITLE
Change "data" to "record"

### LINKED ISSUE #44676: Error when more than one declaration of a type has a parameter list
https://github.com/dotnet/roslyn/blob/891fd421c4488d861c4959bbc91245eb4b118a06/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs#L2926

I assume this refers to a scenario like:
```
partial data class C(int a);
partial data class C(int b);
```

### PR DESCRIPTION
**Pull request by agocke**
Resolves #44676

Adds a new "record" declaration kind and enables when parsing as C#
preview. If C# preview is not enabled, parsing is unchanged.

This PR builds on https://github.com/dotnet/roslyn/pull/44812. Only the last commit is relevant.

This PR takes out the parsing of primary constructors for classes and structs. In retrospect this may have been a mistake since it may have been more work dealing with the broken parsing than semantic errors and fix-ups, but there should be no bad effects aside from diagnostic quality.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -1984,6 +1996,18 @@ private bool IsTypeDeclarationStart()
...
                     return true;
+
+                case SyntaxKind.IdentifierToken:
+                    if (CurrentToken.ContextualKind == SyntaxKind.RecordKeyword)
+                    {
+                        // This is an unusual use of LangVersion. Normally we only produce errors when the langversion
+                        // does not support a feature, but in this case we are effectively making a language breaking
+                        // change to consider "record" a type declaration in all ambiguous cases. To avoid breaking
+                        // older code that is not using C# 9 we conditionally parse based on langversion
+                        return IsFeatureEnabled(MessageID.IDS_FeatureRecords);
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2007]: This is the equivalent of throwing down the gauntlet. I'm excited :smile:

**Comment by gafter** [comment]
Is there a preferred order for reviewing these 544 commits?

**Comment by agocke** [comment]
@gafter 

> This PR builds on #44812. Only the last commit is relevant.

**Comment by agocke** [comment]
I'll reset this PR shortly, when the other one is merge.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs [hunk 2]
```diff
@@ -107,22 +105,68 @@ public void RecordParsing01()
                     }
                     N(SyntaxKind.SemicolonToken);
                 }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+
+            // In langversion 8, this is a method
+            UsingTree(text, options: TestOptions.Regular8,
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:113]: Are there any tests that show the parsing "variants" inside namespaces and/or types? It would be nice if we also get a FeatureInPreview diagnostic there. I'm guessing parsing will just let it through and binding will fail to find a type named 'record'.

I feel like if it's not too difficult to achieve we should try to make sure the workflow of "install VS preview, pop open a C# file in my existing project, and just start using the preview feature" will work. Even something that just notices when a method that returns an ErrorType with the identifier 'record' and adds an additional FeatureInPreview diagnostic seems adequate.

Also, is there any chance that syntax nodes will be reused when changing the LangVersion in an open project? I'm guessing no because today you have to re-parse to get the right diagnostics, but figured it was worth checking.

**Comment by AlekseyTs** [comment]
@agocke Could you re-base this PR?

**Comment by agocke** [comment]
Done

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -1179,22 +1179,7 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
                         break;
 
                     case DeclarationModifiers.Data:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1181]: >case DeclarationModifiers.Data [](start = 20, length = 30)

Do we still need this?

Comment (agocke) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1181]: I'll be using it for the "short property syntax"

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs [hunk 4]
```diff
@@ -24,7 +24,11 @@ internal abstract class SourceConstructorSymbolBase : SourceMemberMethodSymbol
             DiagnosticBag diagnostics)
             : base(containingType, syntax.GetReference(), ImmutableArray.Create(location), SyntaxFacts.HasYieldOperations(syntax))
         {
-            Debug.Assert(syntax.IsKind(SyntaxKind.ConstructorDeclaration) || syntax.IsKind(SyntaxKind.ClassDeclaration) || syntax.IsKind(SyntaxKind.StructDeclaration));
+            Debug.Assert(
+                syntax.IsKind(SyntaxKind.ConstructorDeclaration) ||
+                syntax.IsKind(SyntaxKind.ClassDeclaration) ||
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs:29]: >syntax.IsKind(SyntaxKind.ClassDeclaration) [](start = 16, length = 42)

It feels like we should remove class and struct, i.e. replace them with record. Similarly in other places

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs [hunk 5]
```diff
@@ -16,37 +16,20 @@ internal sealed class SynthesizedRecordConstructor : SourceConstructorSymbolBase
...
-            {
-                case ClassDeclarationSyntax classDecl:
-                    return classDecl.ParameterList!;
-                case StructDeclarationSyntax structDecl:
-                    return structDecl.ParameterList!;
-                default:
-                    throw ExceptionUtilities.Unreachable;
-            }
-        }
+        protected override ParameterListSyntax? GetParameterList() => GetSyntax().ParameterList;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs:32]: >ParameterListSyntax? [](start = 27, length = 20)

Why is this nullable?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs [hunk 6]
```diff
@@ -70,7 +74,9 @@ protected sealed override void MethodChecks(DiagnosticBag diagnostics)
             }
         }
 
-        protected abstract ParameterListSyntax GetParameterList();
+#nullable enable
+        protected abstract ParameterListSyntax? GetParameterList();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs:78]: >ParameterListSyntax? [](start = 27, length = 20)

Why is this nullable?

**Comment by jcouv** [comment]
I'll get started on IDE integration starting from this PR. Not sure all the impact yet, but I expect significant.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs (hunk 6 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs]: That was left over from before the code flow and realized we guarded against null during construction. Fixed.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs (hunk 4 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs]: This was the only place I found, but I can fix up more places if you see them.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs (hunk 2 above)
Comment (agocke) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:113]: It would certainly be useful to get the "FeatureInPreview" diagnostic but I'm going to say that's out-of-scope for this PR.

> Also, is there any chance that syntax nodes will be reused when changing the LangVersion in an open project?

Nope, changing the parse options always requires reparsing every syntax tree from scratch.

**Comment by AlekseyTs** [comment]
>         public static SyntaxKind GetTypeDeclarationKind(SyntaxKind kind)

Should this method be adjusted?

---
Refers to: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:788 in fdb60ac. [](commit_id = fdb60ace2edd2874d231fa4c71f1df7acf9cadc4, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>             for (int i = (int)SyntaxKind.YieldKeyword; i <= (int)SyntaxKind.InitKeyword; i++)

Should this be adjusted?

---
Refers to: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:1074 in fdb60ac. [](commit_id = fdb60ace2edd2874d231fa4c71f1df7acf9cadc4, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs [hunk 7]
```diff
@@ -204,6 +204,8 @@ public enum SyntaxKind : ushort
         DataKeyword = 8441,
         WithKeyword = 8442,
         InitKeyword = 8443,
+        RecordKeyword = 8444,
+
         /// when adding a contextual keyword following functions must be adapted:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs:209]: >when adding a contextual keyword following functions must be adapted: [](start = 12, length = 69)

Please make sure this is taken care of

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs [hunk 8]
```diff
@@ -345,6 +345,7 @@ public static bool IsTypeDeclaration(SyntaxKind kind)
                 case SyntaxKind.InterfaceDeclaration:
                 case SyntaxKind.DelegateDeclaration:
                 case SyntaxKind.EnumDeclaration:
+                case SyntaxKind.RecordDeclaration:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:348]: >case SyntaxKind.RecordDeclaration: [](start = 16, length = 34)

Please make sure we have targeted tests for all modified APIs in this file.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs [hunk 9]
```diff
@@ -980,5 +980,63 @@ public void NominalRecordMembers()
...
+                // (13,17): error CS8863: Only a single record partial declaration may have a parameter list
+                // partial record C(int X, int Y)
+                Diagnostic(ErrorCode.ERR_MultipleRecordParameterLists, "(int X, int Y)").WithLocation(13, 17)
+                );
+
+            Assert.Equal(new[] { "C..ctor(System.Int32 X, System.Int32 Y)", "C..ctor(C )" }, comp.GetTypeByMetadataName("C")!.Constructors.Select(m => m.ToTestDisplayString()));
+        }
+
+        [Fact]
+        public void PartialTypes_02()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs:1014]: >PartialTypes_02 [](start = 20, length = 15)

I think we would want at least one success test. Let's say 3 declarations with initializers and the one in the middle has parameters. Should verify that all members are properly initialized and observe IL to verify order of initialization

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs [hunk 10]
```diff
@@ -73,15 +72,14 @@ class C
...
             UsingTree(text);
 
             N(SyntaxKind.CompilationUnit);
             {
-                N(SyntaxKind.ClassDeclaration);
+                N(SyntaxKind.RecordDeclaration);
                 {
-                    N(SyntaxKind.DataKeyword);
-                    N(SyntaxKind.ClassKeyword);
+                    N(SyntaxKind.IdentifierToken, "record");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:82]: >IdentifierToken [](start = 33, length = 15)

Is this how contextual keywords are surfaced?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 11]
```diff
@@ -1449,25 +1438,25 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
...
                 this.CurrentToken.Kind == SyntaxKind.StructKeyword ||
-                this.CurrentToken.Kind == SyntaxKind.InterfaceKeyword);
+                this.CurrentToken.Kind == SyntaxKind.InterfaceKeyword ||
+                CurrentToken.ContextualKind == SyntaxKind.RecordKeyword);
 
             // "top-level" expressions and statements should never occur inside an asynchronous context
             Debug.Assert(!IsInAsync);
 
-            var classOrStructOrInterface = this.EatToken();
+            var keyword = this.EatToken();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1447]: >var keyword = this.EatToken(); [](start = 12, length = 30)

It looks like this is not going to do the right thing for the "record". Should we use ```EatContextualToken``` for that case?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 12]
```diff
@@ -2669,7 +2693,15 @@ private MemberDeclarationSyntax ParseMemberDeclarationCore(SyntaxKind parentKind
                     MemberDeclarationSyntax result;
                     if (IsMisplacedModifier(modifiers, attributes, type, out result))
                     {
-                        return result;
+                        var misplacedModifier = this.CurrentToken;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2696]: >var misplacedModifier = this.CurrentToken; [](start = 24, length = 42)

It is not clear what is the motivation for this change and why are we not using ```result```.

**Comment by RikkiGibson** [comment]
release_64 failure appears to be a stack size regression.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 12 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Bad merge during rebase. Fixed.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs (hunk 7 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs:209]: Checked, everything seems to be in order.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs (hunk 8 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:348]: I looked around and didn't find any tests for these APIs, it usually looks like we rely on the parsing tests to validate that contextual tokens are properly lexed. Do you want me to add new tests?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 13]
```diff
@@ -1984,6 +1992,18 @@ private bool IsTypeDeclarationStart()
                 case SyntaxKind.InterfaceKeyword:
                 case SyntaxKind.StructKeyword:
                     return true;
+
+                case SyntaxKind.IdentifierToken:
+                    if (CurrentToken.ContextualKind == SyntaxKind.RecordKeyword)
+                    {
+                        // This is an unusual use of LangVersion. Normally we only produce errors when the langversion
+                        // does not support a feature, but in this case we are effectively making a language breaking
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2001]: >making a language breaking [](start = 91, length = 26)

Please add a note in docs for breaking changes

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs [hunk 14]
```diff
@@ -24,7 +24,9 @@ internal abstract class SourceConstructorSymbolBase : SourceMemberMethodSymbol
             DiagnosticBag diagnostics)
             : base(containingType, syntax.GetReference(), ImmutableArray.Create(location), SyntaxFacts.HasYieldOperations(syntax))
         {
-            Debug.Assert(syntax.IsKind(SyntaxKind.ConstructorDeclaration) || syntax.IsKind(SyntaxKind.ClassDeclaration) || syntax.IsKind(SyntaxKind.StructDeclaration));
```
[the remark this answers is not in the document]
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs:27]: I went through all of them and found a bunch more places. I don't want to delay this PR to add tests for all the places, so I added https://github.com/dotnet/roslyn/projects/57#card-39516963 to follow-up

File: src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs [hunk 15]
```diff
@@ -197,13 +197,6 @@ public virtual ImmutableArray<NamedTypeSymbol> GetTypeMembers(string name, int a
             SyntaxKind kind,
             CSharpSyntaxNode syntax)
         {
-            Debug.Assert(
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs:200]: Why remove the assertion?

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs [hunk 16]
```diff
@@ -805,7 +804,7 @@ .maxstack  2
 ");
         }
 
-        [Fact]
+        [Fact(Skip = "record struct")]
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs:807]: nit: link to issue would be better. Or add note to test plan to unskip all tests.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs [hunk 17]
```diff
@@ -980,5 +979,97 @@ public void NominalRecordMembers()
...
+  IL_0008:  ldarg.1
+  IL_0009:  stfld      ""int C.<Y>k__BackingField""
+  IL_000e:  ldarg.0
+  IL_000f:  ldc.i4.2
+  IL_0010:  stfld      ""int C.<Z>k__BackingField""
+  IL_0015:  ldarg.0
+  IL_0016:  call       ""object..ctor()""
+  IL_001b:  ret
+}");
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs:1073]: Let's keep a test for `data class` and maybe even `data record` (both errors)

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs [hunk 18]
```diff
@@ -1189,20 +1313,145 @@ public void WithParsing19()
             EOF();
         }
 
+        [Fact]
+        public void InvalidBase()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:1317]: >InvalidBase [](start = 20, length = 11)

nit: strange test name for this scenario

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs (hunk 8 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:348]: > Do you want me to add new tests?

At least add tests for those that you missed and existing tests didn't catch the fact, or add a note to Test Plan to follow up

---
In reply to: [435439309](https://github.com/dotnet/roslyn/pull/44841#discussion_r435439309) [](ancestors = 435439309)

File: src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs (hunk 15 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs:200]: `ToDeclarationKind` one line below will fire an exception in all instances if the assert isn't followed, so this just feels like duplication

File: docs/compilers/CSharp/Compiler Breaking Changes - post VS2019.md [hunk 19]
```diff
@@ -121,3 +121,14 @@ public class Derived : Base<string?>
 }
 ```
 
+19. https://github.com/dotnet/roslyn/pull/44841 In *C# 9* and onwards the language views ambiguities between the `record` identifier as being
+    either a type syntax or a record declaration as choosing the record declaration. The following examples will now be record declarations:
+
+    ```C#
+    record R1() { }
```
Comment (AlekseyTs) [inline docs/compilers/CSharp/Compiler Breaking Changes - post VS2019.md:128]: >record R1() { } [](start = 4, length = 15)

Is this a valid code in C# 8?

Comment (agocke) [reply docs/compilers/CSharp/Compiler Breaking Changes - post VS2019.md:128]: Good point, it is not.

**Comment by RikkiGibson** [comment]
It appears that DeeplyNestedGeneric is still failing.

**Comment by agocke** [comment]
I'm gonna temporarily reduce the required depth here to get this in. We can fix it up as a bug.
